### PR TITLE
[Tracking] .NET 8 Migration Progress

### PR DESCRIPTION
**Pull request by bagusnl**
[skip ci]
# NOTE: DO NOT MERGE USING GITHUB, WILL BE MERGED MANUALLY LOCALLY 

This PR is to track the changes in preparation for .NET 8 migration.

## Progress
- [x] Initial migration (CollapseLauncher and HI3Helper.Core)
- [x] Submodules migration
   - [x]  ColorThief
   - [x] SevenZipExtractor
   - [x] HI3Helper.SharpHDiffPatch
   - [x] HI3Helper.Http
   - [x] HI3Helper.EncTool
     - [x] UABT
     - [x] CacheStream
- [x] Update dependencies
- [x] Finalize trimmed modules (check which is fine to be trimmed then set it on csproj)
- [x] Code cleanup
- [ ] Bug checks
- [ ] Final code checks

### Bug Fixes
- [x] JSON Handling adjustment
- [x] Memory Allocation issues (particularly when extracting package and applying patch)
- [x] WinAppSDK 1.4
  - [x] ComboBoxItem acrylic BG
  - [ ] ~~PopupFlyout acrylic BG~~ Waiting for fix from MS
  - [x] Random crashes on MainPage nav
- [x] Switch from Reflection based COM to COM Interop via COM Wrappers (FilePicker and SevenZipExtractor)
- [x] Discord RPC keeps getting initialized after being disabled (#304)
- [x] GI HDR Images not getting displayed #296 
> .NET 8 allows interop to be trimmed hence the fix from #298 is not needed anymore

### CODE REVIEW

File: CollapseLauncher/Classes/InstallManagement/StarRail/StarRailInstall.cs
Comment (Cryotechnic) [inline CollapseLauncher/Classes/InstallManagement/StarRail/StarRailInstall.cs:26]: I believe the implementation is the same as GI currently, minus less voicepacks available iirc. Maybe a status update?

File: CollapseLauncher/CollapseLauncher.csproj
Comment (Cryotechnic) [inline CollapseLauncher/CollapseLauncher.csproj:46]: Can we elaborate on what this does exactly?

File: CollapseLauncher/NativeMethods.txt
Comment (Cryotechnic) [inline CollapseLauncher/NativeMethods.txt:1]: Why does this exist?

File: Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs
Comment (Cryotechnic) [inline Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs:171]: Are we sure that we just want to expose application IDs just like that? It allows someone to spoof their application using our ID, which is a potential security risk.

File: Hi3Helper.EncTool.Test/Program.cs
Comment (Cryotechnic) [inline Hi3Helper.EncTool.Test/Program.cs:10]: Genshin EncTool testing :modcheck:

**Review by Cryotechnic — CHANGES_REQUESTED** [review]
Overall looks ok, couple of things to look at, but core implementation seems good

File: CollapseLauncher/Classes/InstallManagement/StarRail/StarRailInstall.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/InstallManagement/StarRail/StarRailInstall.cs:26]: It's "kinda" similar on how GI does it, minus they don't have the pkg_version for the voices and still serve the files via dispatch data too.
Since they are still migrating to this new method, it's kinda hard to say the status of it.

File: CollapseLauncher/CollapseLauncher.csproj
Comment (bagusnl) [reply CollapseLauncher/CollapseLauncher.csproj:46]: This is for launcher patching support where they sometime push an "optional" patch (though sometime its quite critical like HDR bug on HSR recently) through the plugin API (similar to resource API). This has quite complicated implementation because the API doesn't state which version is the served patch for. So it remains experimental for now

File: CollapseLauncher/NativeMethods.txt
Comment (bagusnl) [reply CollapseLauncher/NativeMethods.txt:1]: It's for listing the [NativeMethods that we will use in the project](https://github.com/microsoft/CsWin32#:~:text=Create%20a%20NativeMethods.txt%20file%20in%20your%20project%20directory%20that%20lists%20the%20APIs%20to%20generate%20code%20for.), in this instance CoCreateInstance for the COM used in FileDialogs

File: Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs
Comment (bagusnl) [reply Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs:171]: There isn't a known Discord ApplicationID abuse cases known since its quite useless as it only allows for an RPC showing in your profile and doesn't interact with anything more than that. In fact, some project like PreMID has all their possible ApplicationID exposed as well so I don't believe this is an issue.

File: Hi3Helper.EncTool.Test/Program.cs
Comment (bagusnl) [reply Hi3Helper.EncTool.Test/Program.cs:10]: @neon-nyan is trying to figure it out since code reference used by some repair methods are scattered all over the place. Especially for Genshin as its scattered on 3 projects, main app CollapseLauncher, Core, and EncTool.
This will require moving all the used methods for Genshin repair to EncTool which complicates the submodules and the implementation of it back to the main app. 

So this is a bigger project by itself and require more time that it needs for just an internal testing tool. As of right now, it's fine as is as we hardly ever uses the tester for GI or HI3.

File: .github/workflows/build.yml
Comment (bagusnl) [reply .github/workflows/build.yml:10]: To prevent any issue before merging, this will stay and will get adjusted as the PR is merged

File: CollapseLauncher/Classes/GameManagement/GameVersion/BaseClass/GameVersionBase.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/GameManagement/GameVersion/BaseClass/GameVersionBase.cs:307]: Explained in https://github.com/CollapseLauncher/Collapse/pull/305#discussion_r1398652439

File: CollapseLauncher/Classes/CachesManagement/StarRail/Fetch.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/CachesManagement/StarRail/Fetch.cs:9]: Sure, but it will be always the case since we sort it by alphabet so the `using static CollapseLauncher` will always be on top xD.
Joking aside, since this is just codeQA thing, IMO it's better to stick with just sorting it alphabetically, that way we can use automated CodeQA plugins like VS-internal or ReSharper for sorting all references

File: CollapseLauncher/Classes/FileDialogCOM/FileDialogNative.cs
Comment (neon-nyan) [reply CollapseLauncher/Classes/FileDialogCOM/FileDialogNative.cs:50]: Yes, we will be dropping .NET 7 support entirely and move to .NET 8 since COM Interop is only available on .NET 8 so this pre-define can be removed.
Will do the favor once I'm back home.

File: CollapseLauncher/Classes/FileDialogCOM/FileDialogNative.cs
Comment (neon-nyan) [reply CollapseLauncher/Classes/FileDialogCOM/FileDialogNative.cs:157]: +1 As per comment before

File: CollapseLauncher/Classes/FileDialogCOM/FileDialogNative.cs
Comment (neon-nyan) [reply CollapseLauncher/Classes/FileDialogCOM/FileDialogNative.cs:189]: +2 As per comment before

File: CollapseLauncher/Classes/CachesManagement/StarRail/Fetch.cs
Comment (Cryotechnic) [reply CollapseLauncher/Classes/CachesManagement/StarRail/Fetch.cs:9]: Thanks for clarifying

File: Hi3Helper.EncTool.Test/Program.cs
Comment (Cryotechnic) [reply Hi3Helper.EncTool.Test/Program.cs:10]: That's fair, I believe that we can work on the implementation for Hi3 once 2.0 is closer to release.

File: CollapseLauncher/Classes/InstallManagement/StarRail/StarRailInstall.cs
Comment (Cryotechnic) [reply CollapseLauncher/Classes/InstallManagement/StarRail/StarRailInstall.cs:26]: Is it something that we want to implement for this version or it's still early in development?

Comment (bagusnl) [reply CollapseLauncher/Classes/InstallManagement/StarRail/StarRailInstall.cs:26]: The current implementation should work, though we need to keep an eye on this since it might change on next update

File: CollapseLauncher/Classes/FileDialogCOM/FileDialogNative.cs
Comment (neon-nyan) [reply CollapseLauncher/Classes/FileDialogCOM/FileDialogNative.cs]: As per latest commit, the defines have been deleted

File: CollapseLauncher/NativeMethods.txt
Comment (neon-nyan) [reply CollapseLauncher/NativeMethods.txt]: As per latest commit, I've added the source generated code into App namespace. This file is now unnecessary and has been removed alongside the CsWin32 package.

File: CollapseLauncher/XAMLs/MainApp/MainWindow.xaml
Comment (neon-nyan) [inline CollapseLauncher/XAMLs/MainApp/MainWindow.xaml:12]: @bagusnl Any reason why we added this acrylic effect to the main window? Since we are not using transparent/acrylic effect for the windows backdrop

**Review by neon-nyan — COMMENTED** [review]
Reviewed as per comment

File: CollapseLauncher/XAMLs/MainApp/MainWindow.xaml
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/MainWindow.xaml:12]: Oh, this was leftover from the attempt to fix the acrylic on flyout items, will remove

File: CONTRIBUTING.md
Comment (bagusnl) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md
Comment (bagusnl) [reply CONTRIBUTING.md]: done

**Comment by bagusnl** [comment]
@CollapseLauncher/core-maintainers All conversation that needed change(s) has been resolved. Please do another quick checks before we can merge this PR

**Review by neon-nyan — APPROVED** [review]
I **approb** these changes. LGTM!

Shuba-shuba

**Review by Cryotechnic — APPROVED** [review]
LGTM!

**Comment by bagusnl** [comment]
Approb-al has been received. Merging this PR to main with `deprecated/net-7` branch as backup.
image

**Pull request merged**
